glusterd2/events/eventhandler.go

```diff
@@ -0,0 +1,153 @@
+package events
+
+import (
+	"sort"
+	"strings"
+	"sync"
+)
+
+// Handler defines the event handler interface.
+// It is registered with the events framework to be called when an event
+// happens.
+type Handler interface {
+	// Handle is the function that gets called when an event occurs
+	Handle(*Event)
+	// Events should returns a list of events that the handler is interested in
+	Events() []string
+}
+
+// HandlerID is returned when a Handler is registered. It can be used to unregister a registered Handler.
+type HandlerID uint64
+
+// handler implements the Handler interface around a standalone Handle function
+type handler struct {
+	handle func(*Event)
+	events []string
+}
+
+var (
+	handlers struct {
+		wg sync.WaitGroup
+
+		sync.RWMutex
+		chans map[HandlerID]chan<- *Event
+		next  HandlerID
+	}
+)
+
+func init() {
+	handlers.chans = make(map[HandlerID]chan<- *Event)
+}
+
+func addHandler(c chan<- *Event) HandlerID {
+	handlers.Lock()
+	defer handlers.Unlock()
+
+	id := handlers.next
+	handlers.chans[id] = c
+	handlers.next++
+
+	return id
+}
+
+func delHandler(id HandlerID) chan<- *Event {
+	handlers.Lock()
+	defer handlers.Unlock()
+
+	c, ok := handlers.chans[id]
+	if !ok {
+		return nil
+	}
+	delete(handlers.chans, id)
+	return c
+}
+
+// Register a Handler to be called when the given events happen.
+// If no events are specified, handler function is called for all events
+// Handlers need to be thread safe, as they can be called concurrently when
+// multiple events arrive at the same time.
+func Register(h Handler, events ...string) HandlerID {
```

Comment:
Now that there's `handler.Events()`, you do not need the events list I suppose. Here and in `handleEvents`